File: src/773.cpp
#include <algorithm>
#include <iostream>
#include <queue>
#include <unordered_set>
#include <vector>

using namespace std;

class Solution {
   public:
    int slidingPuzzle(vector<vector<int>>& board) {
        int retVal = 0;

        // Direction map for zero's possible moves in a 1D representation of the 2x3 board
        vector<vector<int>> directions = {{1, 3}, {0, 2, 4}, {1, 5}, {0, 4}, {1, 3, 5}, {2, 4}};
        string target = "123450";

        string startState;
        int boardSize = board.size();
        int boardColSize = board[0].size();
        for (int i = 0; i < boardSize; i++) {  // Convert the 2D board into a string representation
            for (int j = 0; j < boardColSize; j++) {
                startState += to_string(board[i][j]);
            }
        }
        queue<string> bfsQueue;
        bfsQueue.push(startState);
        unordered_set<string> visited;  // To store visited states
        visited.insert(startState);

        // BFS to find the minimum number of moves
        while (bfsQueue.empty() == false) {
            int bfsQueueSize = bfsQueue.size();
            while (bfsQueueSize-- > 0) {
                string currentState = bfsQueue.front();
                bfsQueue.pop();
                if (currentState == target) {  // Check if we reached the target solved state
                    return retVal;
                }

                int zeroPos = currentState.find('0');
                for (int newPos : directions[zeroPos]) {
                    string nextState = currentState;
                    swap(nextState[zeroPos], nextState[newPos]);
                    if (visited.count(nextState)) {  // Skip if this state has been visited
                        continue;
                    }

                    // Mark the new state as visited and add it to the queue
                    bfsQueue.push(nextState);
                    visited.insert(nextState);
                }
            }
            retVal++;
        }
        retVal = -1;

        return retVal;
    }
};

int main(int argc, char** argv) {
    struct subject {
        vector<vector<int>> board;
    };
    vector<subject> testData{{{{1, 2, 3}, {4, 0, 5}}}, {{{1, 2, 3}, {5, 4, 0}}}, {{{4, 1, 2}, {5, 0, 3}}}};
    int numberOfTestCase = testData.size();
    /* Example
     *  Input: board = [[1,2,3],[4,0,5]]
     *  Output: 1
     *
     *  Input: board = [[1,2,3],[5,4,0]]
     *  Output: -1
     *
     *  Input: board = [[4,1,2],[5,0,3]]
     *  Output: 5
     */

    Solution cSolution;
    int answer = 0;
    for (int i = 0; i < numberOfTestCase; ++i) {
        cout << "Input: board = [";
        for (long unsigned int j = 0; j < testData[i].board.size(); ++j) {
            cout << ((j == 0) ? "" : ",") << "[";
            for (long unsigned int k = 0; k < testData[i].board[j].size(); ++k) {
                cout << ((k == 0) ? "" : ",") << testData[i].board[j][k];
            }
            cout << "]";
        }
        cout << "]\n";

        answer = cSolution.slidingPuzzle(testData[i].board);
        cout << "Output: " << answer << "\n";

        cout << "\n";
    }

    return EXIT_SUCCESS;
}

File: src/773.py
import logging
import sys
from bisect import bisect_left, bisect_right
from collections import Counter, defaultdict, deque
from functools import cache, lru_cache
from heapq import heapify, heappop, heappush
from itertools import accumulate
from typing import List, Optional


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


class Solution:
    def swap(self, state, i, j) -> str:
        retVal = ""

        stateList = list(state)
        stateList[i], stateList[j] = stateList[j], stateList[i]
        retVal = "".join(stateList)

        return retVal

    def slidingPuzzle(self, board: List[List[int]]) -> int:
        retVal = 0

        # Direction map for zero's possible moves in a 1D representation of the 2x3 board
        directions = [[1, 3], [0, 2, 4], [1, 5], [0, 4], [1, 3, 5], [2, 4],]
        target = "123450"

        startState = ""
        for rows in board:
            for cell in rows:
                startState += str(cell)
        bfsQueue = deque([startState])
        visited = set()  # To store visited states
        visited.add(startState)

        while bfsQueue:  # BFS to find the minimum number of moves
            queueSize = len(bfsQueue)
            for _ in range(queueSize):
                currentState = bfsQueue.popleft()
                if currentState == target:  # Check if we reached the target solved state
                    return retVal

                zeroPos = currentState.index("0")
                for newPos in directions[zeroPos]:
                    nextState = self.swap(currentState, zeroPos, newPos)
                    if nextState in visited:  # Skip if this state has been visited
                        continue

                    # Mark the new state as visited and add it to the bfsQueue
                    bfsQueue.append(nextState)
                    visited.add(nextState)

            retVal += 1

        retVal = -1

        return retVal


if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        pSolution = Solution()
        for board in [[[1, 2, 3], [4, 0, 5]], [[1, 2, 3], [5, 4, 0]], [[4, 1, 2], [5, 0, 3]]]:
            # /* Example
            #  *  Input: board = [[1,2,3],[4,0,5]]
            #  *  Output: 1
            #  *
            #  *  Input: board = [[1,2,3],[5,4,0]]
            #  *  Output: -1
            #  *
            #  *  Input: board = [[4,1,2],[5,0,3]]
            #  *  Output: 5
            #  */
            logging.info("Input: board = %s", board)

            retVal = pSolution.slidingPuzzle(board)
            logging.info("Output: %s", retVal)

            print()
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass

    sys.exit(0)
